grace/training/config.py

```diff
@@ -96,52 +110,113 @@ def load_config_params(params_file: Union[str, Path]) -> Config:
 
             setattr(config, attr, value)
 
+    return config
+
+
+def validate_required_config_hparams(config: Config) -> None:
+    # Check all required directories are defined:
+    directories = [
+        config.train_image_dir,
+        config.train_grace_dir,
+        config.valid_image_dir,
+        config.valid_grace_dir,
+        config.infer_image_dir,
+        config.infer_grace_dir,
+    ]
+    for dr in directories:
+        if dr is None:
+            raise PathNotDefinedError(path_name=dr)
+        elif not any(dr.iterdir()):
+            raise EmptyDirectoryError(path_name=dr)
         else:
-            continue
+            pass
 
-    return config
+    # Check log_dir exists:
+    if config.log_dir is None:
+        raise PathNotDefinedError(path_name=dr)
+
+    # Check extractor is there:
+    if not config.extractor_fn.is_file():
+        raise PathNotDefinedError(path_name=dr)
+
+    # Define which metrics to calculate:
+    for i in range(len(config.metrics_objects)):
+        m = config.metrics_objects[i].upper()
+        if m == "APPROXIMATE":
+            m = "APPROX"
+        config.metrics_objects[i] = m
+
+    # Make sure saving file suffix is expected:
+    # assert config.saving_file_suffix in {"yaml", "json"}
+
+    # HACK: not automated yet:
+    if config.animate_valid_progress is True:
+        LOGGER.warning("WARNING; auto-animation not implemented yet")
+        config.animate_valid_progress = False
+        # TODO: implemented, but ffmpeg causes issues in tests
+
+    # HACK: not implemented yet:
+    if config.visualise_tsne_manifold is True:
+        LOGGER.warning("WARNING; TSNE manifold not implemented yet")
+        config.visualise_tsne_manifold = False
+        # TODO: implemented, but can't be run from run.py yet
 
 
 def write_config_file(
     config: Config,
     filetype: str = "json",
 ) -> None:
     """Record hyperparameters of a training run."""
-    if filetype not in ["json", "yaml"]:
-        raise ValueError(
-            "Config must be saved as either a .json or .yaml file."
-        )
+    params = {attr: str(getattr(config, attr)) for attr in config.__dict__}
 
-    """params = {}
+    if isinstance(config.run_dir, str):
+        setattr(config, "run_dir", Path(config.run_dir))
 
-    for attr in config.__dict__:
+    fn = config.run_dir / f"config_hyperparams.{filetype}"
+    write_params_as_file_with_suffix(params, fn)
 
-        value = getattr(config, attr)
-        if isinstance(value, function):
-            value = []
 
-        params[attr] = str(value)"""
+def write_params_as_file_with_suffix(
+    parameters_dict: dict[str], filename: str | Path
+) -> None:
+    if isinstance(filename, str):
+        filename = Path(filename)
+
+    if filename.suffix == ".json":
+        with open(filename, "w") as outfile:
+            json.dump(parameters_dict, outfile, indent=4)
+
+    elif filename.suffix == ".yaml":
+        with open(filename, "w") as outfile:
```

Comment:
If I try to run the pipeline starting for the output yaml for a previous run, the code breaks because I think the outputs are changed (e.g lists turned to strings).

This is not the case with the output json file which is consistent.

Reply:
This should now be fixed in the new commit wave! Thank you for the thorough checks - I didn't spot that the params do not parse correctly 🚀